### ISSUE TITLE
Fix lazy evaluation's misleasing effects using the parse tree

### ISSUE THREAD

**Issue by fangly** [rstudio/profvis#79]
Hi,

I stumbled upon a case in which R's lazy evaluation made me think my code was spending a lot of time in a function, when if fact, a totally different function was using my CPU cycles. I found out that this was due to lazy evaluation, as described in the [FAQ](https://rstudio.github.io/profvis/faq.html#why-is-the-call-stack-reversed-for-some-of-my-code).

Here is an example I used:
```R
my_fun <- function() {
  sum(ifelse(runif(10000) > 0.5, 0.8, 0.2) /
      ifelse(runif(10000) > 0.5, 0.1, 0.9))
}

example_1 <- function() {
  for (i in seq_len(100)) {
    print.default(my_fun())
  }
}

example_2 <- function() {
  for (i in seq_len(100)) {
    val <- my_fun()
    print.default(val)
  }
}

library(profvis)
print(profvis(example_1()))  # all time spent used by print (unexpected!)
print(profvis(example_2()))  # all time spent in my_fun (as expected)
```

When profiling, **R's lazy evaluation leads to results that are inaccurate and counter-productive**; how many people will lose time trying to optimize a function that runs fine? The profvis FAQ suggests to use intermediate variables as a workaround, but that assumes that i) the person is aware of the potential issue, and ii) is willing to go through the (perhaps large) codebase and fix all problematic occurrences. Instead, I think profivis should address the issue one way or another:
1. The easiest solution would be to turn off R's lazy evaluation, but it seems there is no way to do that. 
2. Another solution would be to fix the profiling results based on the call tree. In `example_1()` above, which uses lazy evaluation, I have verified using `getParseData(parse("my_example.R"))` that the parse tree is correct, i.e. that `print.default()` and `my_func()` are siblings called by the same parent function, `example_1()`. Thus, the parse tree seems well suited to rectify the profiling results.

Cheers,
Florent

**Comment by wch** [comment]
Unfortunately, the data provided by R's profiler does not include the information needed to do (2). Changing this would require modification to R itself.

**Comment by fangly** [comment]
I am not an Rprof expert, but a quick test reveals that for `example_1()` above, Rprof writes to file call stack lines like:

> "ifelse" "my_fun" "print.default" "example_1" "eval" "eval" "withVisible" "source"
> "runif" "ifelse" "my_fun" "print.default" "example_1" "eval" "eval" "withVisible" "source"
> "is.na" "ifelse" "my_fun" "print.default" "example_1" "eval" "eval" "withVisible" "source"

This seems suitable output for correction. To clarify how I imagined the correction, let me explain in details. Within a `profvis(expr)` call, proceed this way to correct for lazy evaluation:
1. `parsed <- getParseData(parse(text = expr))`. From `parsed`, extract all pairs `c(a, b)` where `a` is the name of a function that was given as argument the result of function `b`. In my example, you would get the pair `c("print.default", "my_fun")`.
2. Run the profiler as usual (simplified): `Rprof("Rprof.out"); force(expr); Rprof(NULL)`
3. Then in `parse_rprof("Rprof.out")` or before, for all pairs identified in step 1, remove function `a` from the call stack lines containing the call 'b a', e.g. replace the call '"my_fun" "print.default"' by '"my_fun"' only.

_Caveat_
There is a problem if `expr` contains a function call lazily-evaluated by another and the same function is also called by the other, e.g.:
```R
print.default <- function(x) {
   my_fun()
   print(x)
}
print.default(my_fun())
```
The profiling file would contain occurrences of '"my_fun" "print.default"', some of which should be corrected, and some others not. Correctly handling this case would require additional information like source file line numbers.

**Comment by wch** [comment]
Removing calls from the stack can cause problems, though. In your example, if the `print.default` function does anything else that takes time, you'd end up with confusing information if you use your method.

Here's an example:

```R
library(profvis)
profvis({
  my_fun <- function() {
    runif(1e6)
  }
  
  f <- function(x) {
    pause(0.1)
    x
    pause(0.1)
  }

  f(my_fun())
})
```

Here's the flame graph:

![image](https://cloud.githubusercontent.com/assets/86978/26115294/2f710760-3a25-11e7-88c2-45752b42b37a.png)

If you replace the samples that have `"f" "myfun"` with just `"myfun"`, then the result would be really confusing. It would look like it exited `f` and then somehow jumped back into the middle of it.

**Comment by fangly** [comment]
Thank you, I appreciate the detailed explanation, Winston. But it seems there is a confusing terminology issue that complicates the discussion...

The plots produced by profvis are [flame _charts_](https://bugs.chromium.org/p/chromium/issues/detail?id=452624), not flame _graphs_, because the x-axis represents the passage of time.

From [Brendan Gregg's description of flame _graphs_](http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html#Description):
> The x-axis spans the sample population. It does not show the passing of time from left to right, as most graphs do. The left to right ordering has no meaning (it's sorted alphabetically to maximize frame merging).

In a flame _graph_, your example would result in a single box for the `pause()` function: 
![prof_expected](https://cloud.githubusercontent.com/assets/231905/26142209/e221dd5c-3adf-11e7-9691-54dd8b6c2d76.png)
And consequently, my suggested correction would result in something like this:
![prof_expected_corrected](https://cloud.githubusercontent.com/assets/231905/26142361/7c79b898-3ae0-11e7-997c-4a21288dd25e.png)
These corrected results very helpful in my opinion, as they better reflect the code.

In short, if profvis generated (real) flame _graphs_, it could also correct them for lazy evaluation.

**Comment by hadley** [comment]
I think we now know how to do this (i.e. in `lobstr::cst()`); we just need to figure out how to patch base R to print the backtrail, rather than the full call tree.

**Comment by lionel-** [comment]
This is now fixed on R 4.0 with https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17595. (Use `simplify = FALSE` to revert to the old output.)
